src/vs/editor/contrib/referenceSearch/browser/referenceSearch.ts

```diff
@@ -7,7 +7,7 @@
 import * as nls from 'vs/nls';
 import { KeyCode, KeyMod } from 'vs/base/common/keyCodes';
 import URI from 'vs/base/common/uri';
-import { TPromise } from 'vs/base/common/winjs.base';
+import { PPromise, TPromise } from 'vs/base/common/winjs.base';
```

Comment:
> What do you mean with an object that the toggle widget can subscribe to?

```
const progressHolder = { onprogress: (value: Location) => void };
const progress = (value: Location) => {
    progressHolder.onprogress(value);
};
const references = provideReferences(model, range.getStartPosition(), progress).then(references => {
    complete(new ReferencesModel(references)), error);
});
controller.toggleWidget(range, references, defaultReferenceSearchOptions, progressHolder);

// Then in toggleWidget

progressHolder.onprogress = (value: Location) => {
    // handle progress
}
```

> Promises with extra sauce are considered an anti pattern by now.

I don't disagree, but the alternative here doesn't seem any better. PPromise is used elsewhere in the VS Code codebase, specifically for streaming search results, so using it here doesn't seem unusual.